bundler/helpers/v2/spec/functions/dependency_source_spec.rb

```diff
@@ -0,0 +1,187 @@
+# frozen_string_literal: true
+
+require "native_spec_helper"
+require "shared_contexts"
+
+RSpec.describe Functions::DependencySource do
+  include_context "in a temporary bundler directory"
+
+  let(:dependency_source) do
+    described_class.new(
+      gemfile_name: "Gemfile",
+      dependency_name: dependency_name
+    )
+  end
+
+  let(:dependency_name) { "business" }
+
+  let(:project_name) { "specified_source_no_lockfile" }
+  let(:registry_url) { "https://repo.fury.io/greysteil/" }
+  let(:gemfury_business_url) do
+    "https://repo.fury.io/greysteil/api/v1/dependencies?gems=business"
+  end
+
+  before do
+    stub_request(:get, registry_url + "versions").
+      with(basic_auth: ["SECRET_CODES", ""]).
+      to_return(status: 404)
+    stub_request(:get, registry_url + "api/v1/dependencies").
+      with(basic_auth: ["SECRET_CODES", ""]).
+      to_return(status: 200)
+    stub_request(:get, gemfury_business_url).
+      with(basic_auth: ["SECRET_CODES", ""]).
+      to_return(status: 200, body: fixture("ruby", "gemfury_response"))
+  end
+
+  describe "#private_registry_versions" do
+    subject(:private_registry_versions) do
+      in_tmp_folder { dependency_source.private_registry_versions }
+    end
+
+    it "returns all versions from the private source" do
+      is_expected.to eq([
+                          Gem::Version.new("1.5.0"),
+                          Gem::Version.new("1.9.0"),
+                          Gem::Version.new("1.10.0.beta")
+                        ])
+    end
+
+    context "specified as the default source" do
+      let(:project_name) { "specified_default_source_no_lockfile" }
+
+      it "returns all versions from the private source" do
+        is_expected.to eq([
+                            Gem::Version.new("1.5.0"),
+                            Gem::Version.new("1.9.0"),
+                            Gem::Version.new("1.10.0.beta")
+                          ])
+      end
+    end
+
+    context "that we don't have authentication details for" do
+      before do
+        stub_request(:get, registry_url + "versions").
+          with(basic_auth: ["SECRET_CODES", ""]).
+          to_return(status: 401)
+        stub_request(:get, registry_url + "api/v1/dependencies").
+          with(basic_auth: ["SECRET_CODES", ""]).
+          to_return(status: 401)
+        stub_request(:get, registry_url + "specs.4.8.gz").
+          with(basic_auth: ["SECRET_CODES", ""]).
+          to_return(status: 401)
+      end
+
+      it "blows up with a useful error" do
+        error_class = Bundler::Fetcher::AuthenticationRequiredError
+        error_message = "Authentication is required for repo.fury.io"
```

Comment:
I have a vague memory of these specs failing because the error_class has changed

Looks like that's handled here https://github.com/dependabot/dependabot-core/blob/cbd62f2611af242457227d49add1f1a88ec38542/bundler/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb#L130

Also tested here https://github.com/dependabot/dependabot-core/blob/cbd62f2611af242457227d49add1f1a88ec38542/bundler/spec/dependabot/bundler/update_checker/latest_version_finder_spec.rb#L286-L297

Kinda annoying that we do regex parsing on the message, which probably means handling the new error outside of the native helper. Maybe one to revisit and refactor, would be neat if the native helper returned a formatted error instead.

Reply:
Yeah, that's odd as it passed for me locally in a container but not on CI. It's failing for me on my native rubyenv now too, will fix.

Reply:
Actually, I think I mispoke, running the tests natively on my Mac seems to result in bundler 1 being used in the v2 native helper 🤔 

I see `spec/dependabot/bundler/update_checker/latest_version_finder_spec.rb` fail two tests on:

```ruby
     Failure/Error:
       expect { finder.latest_version_details }.
         to raise_error do |error|
           expect(error).to be_a Dependabot::GitDependencyReferenceNotFound
           expect(error.dependency).to eq("business")
         end

       expected #<Dependabot::SharedHelpers::HelperSubprocessFailed: private method `URI' called for Bundler:Module> to be a kind of Dependabot::GitDependencyReferenceNotFound
```

I think this is likely working as intended as my native Mac hasn't been provisioned properly to run different bundler versions in a contained way but it is getting kind of tricky to reason about the 'right' test environment.

Reply:
Ahh yes it's a bit of a minefield, I find it works locally if I've installed both, e.g. `gem install bundler:2.2.15 && gem install bundler:1.17.3`

Reply:
I put together a small patch to the native helpers after hitting the above:
https://github.com/dependabot/dependabot-core/pull/3330

I think this is a safe sanity measure _but_ it also proved that the failures in `latest_version_finder_spec.rb` I am seeing locally don't hit this new error, meaning the native helper is using Bundler v2. This failure hasn't turned up in CI as it's blocked by the dependency source failure in the v2 native helpers, so I think there's actually two 'real' bugs.

Reply:
👍🏻  Yeah, it is.

Reply:
Ah but from bundler v1? I changed that monkey patch slightly from the v1 implementation but might still be borked as I only tested it in a dry-run https://github.com/dependabot/dependabot-core/blob/main/bundler/helpers/v1/monkey_patches/git_source_patch.rb#L17

Reply:
👍🏻  I've walked that back in https://github.com/dependabot/dependabot-core/pull/3327/commits/18690c8aaca57ef8e4b1078bb7bf01797941f119 and it passes for me locally

Reply:
Ok, figured out what was going on here - I started using my native environment while waiting for Docker to rebuild which seemed find apart from the failure above. It turns out my native environment had Bundler 2.0.x installed from previous testing, Bundler::URI works in 2.1.x and above.

I'll open a separate PR to update `v2/run.rb` to raise if the Bundler version is lower than 2.1.x since it seems that is the actual API we are dependent on. It might be worth pushing that into the Gemfile itself by depending on bundler but that could get gnarly fast 😬

Reply:
Oh nice find! Yeah sounds like a good idea to check for >= 2.1 👍   nervous of adding `bundler` to the Gemfile, spotted here that dependabot can't upgrade it currently: https://github.com/dependabot/dependabot-core/pull/3319/files#r599508512